LoRaEngine/LoraKeysManagerFacade/ClearLnsCache.cs

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace LoraKeysManagerFacade
+{
+    using System;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using LoRaTools;
+    using Microsoft.AspNetCore.Http;
+    using Microsoft.AspNetCore.Mvc;
+    using Microsoft.Azure.Devices;
+    using Microsoft.Azure.Devices.Common.Exceptions;
+    using Microsoft.Azure.WebJobs;
+    using Microsoft.Azure.WebJobs.Extensions.Http;
+    using Microsoft.Extensions.Logging;
+
+    public class ClearLnsCache
+    {
+        private readonly IEdgeDeviceGetter edgeDeviceGetter;
+        private readonly IServiceClient serviceClient;
+        private readonly IChannelPublisher channelPublisher;
+        private readonly ILogger<ClearLnsCache> logger;
+
+        public ClearLnsCache(IEdgeDeviceGetter edgeDeviceGetter,
+                             IServiceClient serviceClient,
+                             IChannelPublisher channelPublisher,
+                             ILogger<ClearLnsCache> logger)
+        {
+            this.edgeDeviceGetter = edgeDeviceGetter;
+            this.serviceClient = serviceClient;
+            this.channelPublisher = channelPublisher;
+            this.logger = logger;
+        }
+
+        [FunctionName(nameof(ClearNetworkServerCache))]
+        public async Task<IActionResult> ClearNetworkServerCache([HttpTrigger(AuthorizationLevel.Function, "post")] HttpRequest req, CancellationToken cancellationToken)
+        {
+            if (req is null) throw new ArgumentNullException(nameof(req));
+
+            try
+            {
+                VersionValidator.Validate(req);
+            }
+            catch (IncompatibleVersionException ex)
+            {
+                return new BadRequestObjectResult(ex.Message);
+            }
+
+            await ClearLnsCacheInternalAsync(cancellationToken);
+
+            return new AcceptedResult();
+        }
+
+        internal async Task ClearLnsCacheInternalAsync(CancellationToken cancellationToken)
+        {
+            this.logger.LogInformation("Clearing device cache for all edge and Pub/Sub channel based Network Servers.");
+            var edgeDevices = await this.edgeDeviceGetter.ListEdgeDevicesAsync(cancellationToken);
+            var tasks = edgeDevices.Select(e => InvokeClearViaDirectMethodAsync(e, cancellationToken)).ToArray();
+
+            await PublishClearMessageAsync();
+            await Task.WhenAll(tasks);
```

Comment:
If the synchronous part of a `InvokeClearViaDirectMethodAsync` fails (currently the first operation is already async, so it's not yet a problem) we do not publish the message on Redis. Consider yielding the tasks initially to make it explicit that they will always attempt to execute no matter what.